nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp

```diff
@@ -0,0 +1,268 @@
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "BinaryFilter: New costmap filter info arrived from %s topic. Updating old filter info.",
+      filter_info_topic_.c_str());
+    // Resetting previous subscriber each time when new costmap filter information arrives
+    mask_sub_.reset();
+  }
+
+  if (msg->type != BINARY_FILTER) {
+    RCLCPP_ERROR(logger_, "BinaryFilter: Mode %i is not supported", msg->type);
+    return;
+  }
+
+  // Set base_ and multiplier_
+  base_ = msg->base;
+  multiplier_ = msg->multiplier;
+  // Set topic name to receive filter mask from
+  mask_topic_ = msg->filter_mask_topic;
+
+  // Setting new filter mask subscriber
+  RCLCPP_INFO(
+    logger_,
+    "BinaryFilter: Subscribing to \"%s\" topic for filter mask...",
+    mask_topic_.c_str());
+  mask_sub_ = node->create_subscription<nav_msgs::msg::OccupancyGrid>(
+    mask_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&BinaryFilter::maskCallback, this, std::placeholders::_1));
+}
+
+void BinaryFilter::maskCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  if (!filter_mask_) {
+    RCLCPP_INFO(
+      logger_,
+      "BinaryFilter: Received filter mask from %s topic.", mask_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "BinaryFilter: New filter mask arrived from %s topic. Updating old filter mask.",
+      mask_topic_.c_str());
+    filter_mask_.reset();
+  }
+
+  filter_mask_ = msg;
+  mask_frame_ = msg->header.frame_id;
+}
+
+void BinaryFilter::process(
+  nav2_costmap_2d::Costmap2D & /*master_grid*/,
+  int /*min_i*/, int /*min_j*/, int /*max_i*/, int /*max_j*/,
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  if (!filter_mask_) {
+    // Show warning message every 2 seconds to not litter an output
+    RCLCPP_WARN_THROTTLE(
+      logger_, *(clock_), 2000,
+      "BinaryFilter: Filter mask was not received");
+    return;
+  }
+
+  geometry_msgs::msg::Pose2D mask_pose;  // robot coordinates in mask frame
+
+  // Transforming robot pose from current layer frame to mask frame
+  if (!transformPose(global_frame_, pose, mask_frame_, mask_pose)) {
+    return;
+  }
+
+  // Converting mask_pose robot position to filter_mask_ indexes (mask_robot_i, mask_robot_j)
+  unsigned int mask_robot_i, mask_robot_j;
+  if (!worldToMask(filter_mask_, mask_pose.x, mask_pose.y, mask_robot_i, mask_robot_j)) {
+    // Robot went out of mask range. Set "false" state by-default
+    RCLCPP_WARN(
+      logger_,
+      "BinaryFilter: Robot is outside of filter mask. Resetting binary state to default.");
+    changeState(default_state_);
+    return;
+  }
+
+  // Getting filter_mask data from cell where the robot placed
+  int8_t mask_data = getMaskData(filter_mask_, mask_robot_i, mask_robot_j);
+  if (mask_data == nav2_util::OCC_GRID_UNKNOWN) {
```

Comment:
Remove, the unknown costs are only relevant for cost values, this is not that :smile:

Reply:
The `mask_data` does not belong to Costmap2D space, it is `OccupancyGrid` value.
OccGrid values  are in the range from `[0,100]`, or unknown which is `-1`. This is exception value we do not want to decode back to the filter space value, therefore this situation should be pulled-out and handled separately.

Reply:
The similar case is already [handled](https://github.com/ros-planning/navigation2/blob/main/nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp#L276) in `SpeedFilter`, but for some reasons (I can not recall why), we are using separate const value names for it.

Reply:
OK. I didn't recall that.